### ISSUE TITLE
Set the weights in tf.layers with other variables but not as initializers.

### ISSUE THREAD

**Issue by siavash-khodadadeh** [tensorflow/tensorflow#19108]
### System information
- **Have I written custom code**: N/A (There is no example of meta learning with tf.layers)
- **OS Platform and Distribution**: Linux Ubuntu 16.04
- **TensorFlow installed from**: Installed tf-nightly with pip 
- **TensorFlow version**: v1.8.0-rc1-934-g291d85be42 1.9.0-dev20180426
- **Python version**: 3.5.2 
- **Bazel version**: N/A
- **GCC/Compiler version**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: N/A

### Describe the problem
Nowadays, with advances in deep learning, the researchers sometimes need to set the weights of their model based on some formula (not initialize them but initialize something which calculates that formula) and then use that to backpropagate through the variables which computed that formula. For example in meta learning, I have a model, let's say it is just one layer tf.layers.dense(). I want to compute gradients and then use that gradients to build updated model graph and compute gradients on that. The backpropagation should tell me how should I update my first model weights to adapt well. Look at the code below:

### Source code / logs

    train = tf.placeholder(dtype=tf.float32, shape=(None, 4), name='train')
    train_out = tf.placeholder(dtype=tf.float32, shape=(None, 1), name='train_out')

    validation = tf.placeholder(dtype=tf.float32, shape=(None, 4))
    validation_out = tf.placeholder(dtype=tf.float32, shape=(None, 1))

    with tf.variable_scope('model'):
        model_out = tf.layers.dense(train, 1, activation=tf.nn.relu)

    with tf.variable_scope('loss'):
        loss = tf.square(model_out - train_out)

    with tf.variable_scope('gradients'):
        optimizer = tf.train.AdamOptimizer()
        grads = optimizer.compute_gradients(loss)

    with tf.variable_scope('updated_model'):
        updated_vars = {
            grad_info[1].name[6:]: grad_info[1] - 0.1 * grad_info[0] \
            for grad_info in grads if grad_info[0] is not None
            }

        # meta_out = tf.nn.relu(tf.matmul(validation, updated_vars['dense/kernel:0']) + updated_vars['dense/bias:0'])
        meta_out = tf.layers.dense(validation, 1, activation=tf.nn.relu, 
    kernel_initializer=updated_vars['dense/kernel:0'], bias_initializer=updated_vars['dense/bias:0'])

    with tf.variable_scope('meta_loss'):
        meta_loss = tf.square(meta_out - validation_out)

    with tf.variable_scope('meta_optimizer'):
        model_vars = tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, scope='model')
        meta_optimizer = tf.train.AdamOptimizer()
        meta_train_op = meta_optimizer.minimize(meta_loss, var_list=model_vars)

In this code, I have commented one line. In that line, I did not use tf.layers but instead implemented what should happen in that layer myself. I think we need to be able to do that with tf.layers as well. 
Please notice that we cannot use initializer because initializers should be set when we run tf.global_variables_initializer() and have another meaning(Which is to set the value of a variable). Here we do not want to initial those weights of layers with values but just with variables which we calculated in some other way and be able to backpropagate through them.

I think one way to solve this is to allow creating tf.layers' models with setting kernel and bias directly instead of using initializers. I would be more than happy if I could help with this issue.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by siavash-khodadadeh** [comment]
Sorry about that. I think I missed the form. I have completed the required information now.

**Comment by karmel** [comment]
@fchollet , thoughts on returning more information from layers as requested here?

**Comment by fchollet** [comment]
So you want the ability to set layer's weights as custom tensors?

When we talk about "layer's weights", there are two components:

- Layer attributes such as `self.kernel`, `self.bias`.
- List of weights that are used in backpropagation (`self.trainable_weights`) or used for saving (`self.weights`).

It seems you want to:

1) Explicitly create your own weight variables
2) Transform them through a formula
3) Use the output of that as the `self.kernel` (for instance) attribute on your layer
4) Still be able to update the original variables that you created, via backprop when training your model

Is that right?

It sounds like you should subclass the layer you're looking at (e.g. `Dense`) and write your own `build` methods (the method that creates the weights).

**Comment by siavash-khodadadeh** [comment]
Yes. That is exactly what I wanted. Maybe that is a better workaround to subclass them. I also did almost the same thing to solve my problem. However, it seems kind of intuitive to me that I should be able to create those layers with whatever variable I want as weights and bias. As a result, I do not need to go to lower-level API whenever I want to do something like this and can use these higher-level APIs easily. What are your thoughts about this?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 21 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 37 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 52 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 67 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 82 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 97 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yashkant** [comment]
Hi, 

I have built a wrapper around tf.layers.Conv2D. I want to define the layer weights explicitly in the graph as a tensor and then add it to the layer. I want to use this variable multiple times during the training phase. 

Something like: 
`
layer_weights = tf.get_variable(name,shape)
x = my_custom_layer(x, weights, args..) 

`
I want to know what should I write in the build function of the custom layer to set the layer weights to a pre-defined tensor. Please Help.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 90 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @fchollet: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mztkenan** [comment]
The same problem when implement LambdaOpt，to set layer's weights as custom tensors.

**Comment by sushreebarsa** [comment]
@siavash-khodadadeh In TensorFlow 2.x, there is no placeholder. Please take a look at the [official doc](https://www.tensorflow.org/alpha/guide/upgrade) on converting your TF1.x code to TF2.x. and try with the latest TF version 2.9.0 as older versions are not actively supported.
Thank you!
